keps/sig-node/4381-dra-semantic-parameters/README.md

```diff
@@ -0,0 +1,1477 @@
+the framework as a whole. This model is a natural extension to the API provided
+by the traditional device-plugin, which supports advertising resources as a
+finite set of opaque strings. The major difference being that under the new
+model, one can attach a list of attributes to those opaque strings which the
+scheduler can use to select a particular resource for allocation. Additionally,
+each resource can be divided into a set of pre-defined partitions, each with
+their own set of unique attributes attached to them. In this way, requests for
+resources can be satisfied either by a complete resource instance or by one of
+its partitions.  This model may get enhanced in the future and/or other models
+might get added as needed.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+- Enable cluster autoscaling when pods use resource claims, with correct
+  decisions and changing the cluster size by more than one node at a time.
+
+- Support node-local resources. Adding or removing nodes has no effect
+  on network-attached resources and therefore CA does not need to (and cannot)
+  simulate them.
+
+- Allow DRA driver developers to provide a user experience that is similar to
+  the one possible without semantic parameters. Ideally, users should not notice
+  at all that a driver is using semantic parameters under the hood.
+
+### Non-Goals
+
+- Scheduling performance is expected to become better compared to using the
+  PodSchedulingContext. However, this is not the reason for this KEP.
+
+
+## Proposal
+
+### Publishing node resources
+
+The kubelet publishes NodeResourceSlices to the API server with content
+provided by DRA drivers running on each node. Access control through the node
+authorizer ensures that the kubelet running on one node is not allowed to
+create or modify NodeResourceSlices belonging to other nodes.
+
+NodeResourceSlices are published separately for each driver, using a version of
+the API that matches what that driver is using. This implies that the kubelet, DRA
+driver, and apiserver must all support the *same* API version. It might be
+possible to support version skew (= keeping kubelet at an older version than
+the control plane and the DRA drivers) in the future, but currently this is out
+of scope.
+
+Embedded inside a NodeResourceSlice is the representation of the resources
+managed by a driver. With the "semantic model" of `PartitionableResources` this
+takes the form:
+
+```yaml
+kind: NodeResourceSlice
+apiVersion: resource.k8s.io/v1alpha2
+...
+spec:
+  nodeName: worker-1
+  driverName: cards.dra.example.com
+  partitionable:
```

Comment:
I assume that since this is still a `Slice`, then in theory the same driver can publish multiple of these for the same node, correct? The scheduler would need to concatenate those all internally - or be able to evaluate them all for a given node.

Reply:
It's possible in theory, but what is the use case you are thinking of where a driver might want to do this? I guess by the sheer fact that it is possible we would need machinery to do such a concatenation in case a driver did decide to do this.

Reply:
Too big for etcd would be the only reason (ie, necessity not desire)

Reply:
The content of the semantic model type determines whether slicing it up is possible. I'll add a comment to that regard.

In the current model, this would be easy at the top `resourceInstances` array. Ordering the search then might be an issue, so this would need further explanation at that level.

Reply:
> Whether the information about resources must fit into one NodeResourceSlice
> object or can be distributed across many depends on the semantic model and how
> it describes resources. The size of each object is a hard limit that must be
> taken into account when designing a semantic model and preparing
> NodeResourceSlice objects.